### PR TITLE
feat: add `@stdlib/blas/base/drotg`

### LINKED ISSUE #279: RFC: add BLAS Level 1 drotg routine
## Checklist

> Please ensure the following tasks are completed before submitting a feature request.

-   [X] Read and understood the [Code of Conduct][code-of-conduct].
-   [X] Searched for existing issues and pull requests.
-   [X] The issue name begins with `RFC:`.

## Description

> Description of the feature request.

This RFC proposes to add a routine to construct a Givens plane rotation  as defined in [BLAS Level 1][blas-level-1] routines.  Specifically adding `@stdlib/blas/base/drotg` is proposed.

## Related Issues

> Does this feature request have any related issues?

No.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this feature request? This may include screenshots, references, sample output, and/or implementation notes.

No.

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[blas-level-1]: http://www.netlib.org/blas/#_level_1

### LINKED ISSUE #280: RFC: add BLAS Level 1 srotg routine
## Checklist

> Please ensure the following tasks are completed before submitting a feature request.

-   [X] Read and understood the [Code of Conduct][code-of-conduct].
-   [X] Searched for existing issues and pull requests.
-   [X] The issue name begins with `RFC:`.

## Description

> Description of the feature request.

This RFC proposes to add a routine to construct a Givens plane rotation as defined in [BLAS Level 1][blas-level-1] routines.  Specifically adding `@stdlib/blas/base/srotg` is proposed.

## Related Issues

> Does this feature request have any related issues?

No.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this feature request? This may include screenshots, references, sample output, and/or implementation notes.

No.

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[blas-level-1]: http://www.netlib.org/blas/#_level_1

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #279.

## Checklist

- [x] Add readme.md
- [x] repl.txt
- [x] types
- [x] package.json
- [x] lib
- [x] examples
- [x] benchmark
- [x] test 
* * *

@kgryte

----

Reference `drotg` implementation: [netlib](https://www.netlib.org/lapack/explore-html/d7/dc5/group__rotg_gaafa91c51f75df6c3f2182032a221c2db.html#gaafa91c51f75df6c3f2182032a221c2db)

### CODE REVIEW

**Comment by Pranavchiku** [comment]
I guess, I have a lint error in `test.assign.js` and `test.main.js`, that was some `mixed space and tab indents`, I was unable to fix it hence used `--no-verify` commit, will polish it once all files get in this PR.

**Comment by Pranavchiku** [comment]
All tests and benchmarks cleared, this PR can have a review.

**Comment by kgryte** [comment]
@Pranavchiku Would you be willing to update this PR according to the changes made in commit <https://github.com/stdlib-js/stdlib/commit/36b75b3271e9c3ee29c2c4488c3c1420fe48dd2e>?

**Comment by kgryte** [comment]
@Pranavchiku When getting mixed tabs and spaces lint errors, in Sublime Text, I simply select all text and then use built in IDE tools to convert indentation from spaces to tabs (or vice versa depending on the file type). In short, prefer resolving this locally rather than using `--no-verify`.

File: lib/node_modules/@stdlib/blas/base/drotg/benchmark/benchmark.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+*/
+
+'use strict';
+
+// MODULES //
+
+var bench = require( '@stdlib/bench' );
+var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var Float32Array = require( '@stdlib/array/float32' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/benchmark/benchmark.js]:
```suggestion
var Float64Array = require( '@stdlib/array/float64' );
```

File: lib/node_modules/@stdlib/blas/base/drotg/benchmark/benchmark.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+bench( pkg+':assign', function benchmark( b ) {
+	var out;
+	var x;
+	var y;
+	var z;
+	var i;
+
+	x = discreteUniform( 100, -5, 5, OPTS );
+	y = discreteUniform( 100, -5, 5, OPTS );
+	out = new Float32Array( 4 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/benchmark/benchmark.js]:
```suggestion
	out = new Float64Array( 4 );
```

File: lib/node_modules/@stdlib/blas/base/drotg/lib/assign.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var abs = require( '@stdlib/math/base/special/abs' );
+var sqrt = require( '@stdlib/math/base/special/sqrt' );
+var pow = require( '@stdlib/math/base/special/pow' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/lib/assign.js]:
```suggestion
var abs2 = require( '@stdlib/math/base/special/abs2' );
```

File: lib/node_modules/@stdlib/blas/base/drotg/lib/assign.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		sign = copysign( 1.0, b );
+	}
+	scale = aa + ab;
+	if ( scale === 0.0 ) {
+		c = 1.0;
+		s = 0.0;
+		r = 0.0;
+		z = 0.0;
+	} else {
+		r = scale * sqrt( pow( a/scale, 2 ) + pow( b/scale, 2 ) );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/lib/assign.js]:
```suggestion
		r = scale * sqrt( abs2( a/scale ) + abs2( b/scale ) );
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+		drotg.assign( values[i][0], values[i][1], out, 1, 0 );
+		if ( out === expected[i] ) {
+			t.deepEqual( out, expected[i], 'returns expected value' );
+		} else {
+			for ( j = 0; j < out.length; j++ ) {
+				delta = abs( out[j] - expected[i][j] );
+				tol = 1.5 * EPS * abs( expected[i][j] );
+				t.ok( delta <= tol, 'within tolerance. out: '+out[j]+'. expected: '+expected[i][j]+'. delta: '+delta+'. tol: '+tol+'.' );
+			}
+		}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]:
```suggestion
		for ( j = 0; j < out.length; j++ ) {
			if ( out[j] === e[j] ) {
				t.strictEqual( out[j], e[j], 'returns expected value' );
			} else {
				delta = abs( out[j] - e[j] );
				tol = 1.5 * EPS * abs( e[j] );
				t.ok( delta <= tol, 'within tolerance. out: '+out[j]+'. expected: '+e[j]+'. delta: '+delta+'. tol: '+tol+'.' );
			}
		}
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	}
+	if ( actual === expected ) {
+		t.deepEqual( actual, expected, 'returns expected value' );
+	} else {
+		for ( i = 0; i < actual.length; i++ ) {
+			delta = abs( actual[i] - expected[i] );
+			tol = 1.5 * EPS * abs( expected[i] );
+			t.ok( delta <= tol, 'within tolerance. actual: '+actual[i]+'. expected: '+expected[i]+'. delta: '+delta+'. tol: '+tol+'.' );
+		}
+	}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]: These tests do not work as expected. The `out === expected` comparison is comparing references, not values. These tests and below need to be refactored.

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+	t.strictEqual( drotg.assign.length, 5, 'arity of 5' );
+	t.end();
+});
+
+tape( 'the function computes a Givens plane rotation', function test( t ) {
+	var expected;
+	var values;
+	var delta;
+	var tol;
+	var out;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js:49]:
```suggestion
	var out;
	var e;
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+		[ -0.3, 0.4 ],
+		[ -0.4, 0.3 ],
+		[ -0.3, -0.4 ],
+		[ 0.0, 0.0 ],
+		[ 0.0, 1.0 ],
+		[ 1.0, 0.0 ]
+	];
+
+	for ( i = 0; i < values.length; i++ ) {
+		expected[i] = new Float64Array( expected[i] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]:
```suggestion
		e = new Float64Array( expected[i] );
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js (hunk 6 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]: See refactoring above.

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.main.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+		[ -0.3, 0.4 ],
+		[ -0.4, 0.3 ],
+		[ -0.3, -0.4 ],
+		[ 0.0, 0.0 ],
+		[ 0.0, 1.0 ],
+		[ 1.0, 0.0 ]
+	];
+
+	for ( i = 0; i < values.length; i++ ) {
+		expected[i] = new Float64Array( expected[i] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.main.js]: Same comments as `test.assign.js`. These tests need to be refactored.

**Review by kgryte — CHANGES_REQUESTED** [review]
@Pranavchiku This mostly looks good. The tests need some refactoring.

**Comment by Pranavchiku** [comment]
Updated tests, all tests pass and benchmarks cleared, this PR can have a final review.

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+	}
+	for ( i = 0; i < out.length; i++ ) {
+		if ( out[i] === actual[i] ) {
+			t.strictEqual( out[i], actual[i], 'returns expected value' );
+		} else {
+			delta = abs( out[i] - actual[i] );
+			tol = 1.5 * EPS * abs( actual[i] );
+			t.ok( delta <= tol, 'within tolerance. out: '+out[i]+'. expected: '+actual[i]+'. delta: '+delta+'. tol: '+tol+'.' );
+		}
+	}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]:
```suggestion
```

This appears to be duplicated.

Comment (kgryte) [reply lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]: Ah. Okay, I see what you are doing here. This is unnecessary. We just need to test whether `actual === out`.

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+	var delta;
+	var tol;
+	var out;
+	var i;
+
+	expected = new Float64Array( [ 0.5, 0.0, 1.0 / 0.6, 0.0, 0.6, 0.0, 0.8 ] );
+	out = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );
+
+	actual = drotg.assign( 0.3, 0.4, out, 2, 0 );
+	for ( i = 0; i < out.length; i++ ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js:126]:
```suggestion
	t.strictEqual( actual, out, 'returns expected value' );
	for ( i = 0; i < out.length; i++ ) {
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+	var delta;
+	var tol;
+	var out;
+	var i;
+
+	expected = new Float64Array( [ 0.8, 0.0, 0.6, 0.0, 1.0 / 0.6, 0.0, 0.5 ] );
+	out = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );
+
+	actual = drotg.assign( 0.3, 0.4, out, -2, 6 );
+	for ( i = 0; i < out.length; i++ ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js:151]:
```suggestion
	t.strictEqual( actual, out, 'returns expected value' );
	for ( i = 0; i < out.length; i++ ) {
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+	}
+	for ( i = 0; i < out.length; i++ ) {
+		if ( out[i] === actual[i] ) {
+			t.strictEqual( out[i], actual[i], 'returns expected value' );
+		} else {
+			delta = abs( out[i] - actual[i] );
+			tol = 1.5 * EPS * abs( actual[i] );
+			t.ok( delta <= tol, 'within tolerance. out: '+out[i]+'. expected: '+actual[i]+'. delta: '+delta+'. tol: '+tol+'.' );
+		}
+	}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+	var delta;
+	var tol;
+	var out;
+	var i;
+
+	expected = new Float64Array( [ 0.0, 0.5, 1.0 / 0.6, 0.6, 0.8 ] );
+	out = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0 ] );
+
+	actual = drotg.assign( 0.3, 0.4, out, 1, 1 );
+	for ( i = 0; i < out.length; i++ ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js:176]:
```suggestion
	t.strictEqual( actual, out, 'returns expected value' );
	for ( i = 0; i < out.length; i++ ) {
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+	}
+	for ( i = 0; i < out.length; i++ ) {
+		if ( out[i] === actual[i] ) {
+			t.strictEqual( out[i], actual[i], 'returns expected value' );
+		} else {
+			delta = abs( out[i] - actual[i] );
+			tol = 1.5 * EPS * abs( actual[i] );
+			t.ok( delta <= tol, 'within tolerance. out: '+out[i]+'. expected: '+actual[i]+'. delta: '+delta+'. tol: '+tol+'.' );
+		}
+	}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+	var delta;
+	var tol;
+	var out;
+	var i;
+
+	expected = new Float64Array( [ 0.0, 0.0, 0.5, 0.0, 1.0 / 0.6, 0.0, 0.6, 0.0, 0.8 ] ); // eslint-disable-line max-len
+	out = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );
+
+	actual = drotg.assign( 0.3, 0.4, out, 2, 2 );
+	for ( i = 0; i < out.length; i++ ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js:201]:
```suggestion
	t.strictEqual( actual, out, 'returns expected value' );
	for ( i = 0; i < out.length; i++ ) {
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+	}
+	for ( i = 0; i < out.length; i++ ) {
+		if ( out[i] === actual[i] ) {
+			t.strictEqual( out[i], actual[i], 'returns expected value' );
+		} else {
+			delta = abs( out[i] - actual[i] );
+			tol = 1.5 * EPS * abs( actual[i] );
+			t.ok( delta <= tol, 'within tolerance. out: '+out[i]+'. expected: '+actual[i]+'. delta: '+delta+'. tol: '+tol+'.' );
+		}
+	}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]:
```suggestion
```

File: lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+		} else {
+			delta = abs( out[i] - expected[i] );
+			tol = 1.5 * EPS * abs( expected[i] );
+			t.ok( delta <= tol, 'within tolerance. out: '+out[i]+'. expected: '+expected[i]+'. delta: '+delta+'. tol: '+tol+'.' );
+		}
+	}
+	t.end();
+});
+
+tape( 'the function supports providing both stride and offset', function test(t) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/drotg/test/test.assign.js]:
```suggestion
tape( 'the function supports providing both a stride and offset', function test(t) {
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @Pranavchiku!

**Pull request merged**